File: src/main/kotlin/frc/team2186/robot/autonomous/CommonAuto.java
package frc.team2186.robot.autonomous;

import frc.team2186.robot.Config;
import frc.team2186.robot.lib.common.IterativeAutoAction;
import frc.team2186.robot.lib.math.Rotation2D;
import frc.team2186.robot.subsystems.Drive;
import frc.team2186.robot.subsystems.Platform;
import kotlin.Unit;
import kotlin.jvm.functions.Function0;

public class CommonAuto {
    public static IterativeAutoAction goDistance(double distance) {
        return new IterativeAutoAction(() -> {
            Drive.INSTANCE.setForwardVelocity(Config.Auto.speed);
            return Drive.INSTANCE.getLeftPosition() > distance;
        });
    }

    public static IterativeAutoAction turnToAngle(double angle) {
        return new IterativeAutoAction(() -> {
            Drive.INSTANCE.setGyroSetpoint(Rotation2D.Companion.fromDegrees(angle));

            return Drive.INSTANCE.getGyroAngle().getDegrees() == angle;
        });
    }

    public static Function0<Unit> stop() {
        return () -> {
            Drive.INSTANCE.stop();
            return Unit.INSTANCE;
        };
    }

    public static IterativeAutoAction baseline(boolean half) {
        return half ? goDistance(103 / 2) : goDistance(103);
    }

    public static IterativeAutoAction crossBoard(boolean half) {
        return half ? goDistance(137.5 / 2) : goDistance(137.5);
    }

    public static IterativeAutoAction heading90() {
        return turnToAngle(90);
    }

    public static IterativeAutoAction heading270() {
        return turnToAngle(270);
    }

    public static IterativeAutoAction dropBox() {
        return new IterativeAutoAction(() -> {
            Platform.INSTANCE.setSetpoint(0.25);

            if (Platform.INSTANCE.isBarUp()) {
                Platform.INSTANCE.setSetpoint(-0.25);
                return Platform.INSTANCE.isBarDown();
            }
            return false;
        });
    }

    public static IterativeAutoAction moveOne() {
        return goDistance(80);
    }

    public static IterativeAutoAction moveTwo() {
        return goDistance(59);
    }
}

File: src/main/kotlin/frc/team2186/robot/autonomous/CommonAuto.kt
package frc.team2186.robot.autonomous

import frc.team2186.robot.Config
import frc.team2186.robot.lib.common.withinRange
import frc.team2186.robot.subsystems.Drive
import frc.team2186.robot.subsystems.Platform

/*
object CommonAuto {
    fun goDistance(distance: Double): IterativeAutoAction {
        return IterativeAutoAction {
            OldDrive.setForwardVelocity(Config.Auto.speed)
            OldDrive.leftPosition > distance
        }
    }

    fun turnToAngle(angle: Double): IterativeAutoAction {
        return IterativeAutoAction {
            OldDrive.gyroSetpoint = Rotation2D.fromDegrees(angle)

            OldDrive.gyroAngle.degrees == angle
        }
    }

    fun stop(): Function0<Unit> {
        return {
            OldDrive.stop()
            Unit
        }
    }

    fun baseline(half: Boolean): IterativeAutoAction {
        return if (half) goDistance((103 / 2).toDouble()) else goDistance(103.0)
    }

    fun crossBoard(half: Boolean): IterativeAutoAction {
        return if (half) goDistance(137.5 / 2) else goDistance(137.5)
    }

    fun heading90(): IterativeAutoAction {
        return turnToAngle(90.0)
    }

    fun heading270(): IterativeAutoAction {
        return turnToAngle(270.0)
    }

    fun dropBox(): IterativeAutoAction {
        return IterativeAutoAction {
            Platform.setpoint = 0.25

            if (Platform.isBarUp) {
                Platform.setpoint = -0.25
                Platform.isBarDown
            }
            false
        }
    }

    fun moveOne(): IterativeAutoAction {
        return goDistance(80.0)
    }

    fun moveTwo(): IterativeAutoAction {
        return goDistance(59.0)
    }
}
*/

fun distance(length: Double): () -> Boolean {
    return {
        Drive.setForwardVelocity(Config.Auto.speed)
        Drive.leftPosition >= length
    }
}

fun turnToAngle(angle: Double): () -> Boolean {
    return {
        Drive.setVelocityVector(0.0, angle)
        withinRange(Drive.heading, angle - 1, angle + 1)
    }
}

fun baseline(half: Boolean) = distance(85.0 / (if (half) 2 else 1))
fun cross(half: Boolean) = distance(137.5 / (if (half) 2 else 1))
fun heading90() = turnToAngle(90.0)
fun heading270() = turnToAngle(270.0)
fun moveOne() = distance(40.0)
fun moveTwo() = distance(45.0)
fun dropBox() = {
    Platform.setpoint = -0.5
    if (Platform.isBarUp) {
        Platform.setpoint = 0.0
        true
    }
    false
}